Load vendor configuration directories outside install prefix

krobelus opened the pull request:
Vendor configuration provided by third party packages is sourced first
from `$EXTRA_VENDOR_PATH/vendor_*` and `$PREFIX/share/fish/vendor_*`
instead of only the latter.  The default for `$EXTRA_VENDOR_PATH` is
`/usr/local/share/fish:/usr/share/fish`, which enables fish to source
third party configuration from packages installed there, even if fish is
installed with a different prefix. The directories in `$EXTRA_VENDOR_PATH`
will not be created by (installing) fish.

In particular, with this change
`/usr/bin/fish` also uses `/usr/local/share/fish/vendor_completions.d`, and
`/usr/local/bin/fish` also uses `/usr/share/fish/vendor_completions.d`

Fixes #5029

## Description

Not sure if this is right, I haven't tested this on different systems.

But it does solve issues like missing upstreams completions for say `docker` when the user installs `docker` to `/usr` and fish to `usr/local`, or vice versa. Which seem like legitimate uses.
It also allows a fish install to a random `$PREFIX` to work just as well installing it to `/usr/local`, which is nice for development.

I guess distros like Arch that don't have `/usr/local/bin` in the default `$PATH` would want to set `EXTRA_VENDOR_PATH=""`, but I think the default of `/usr/local/share/fish:/usr/share/fish` is safe for most systems out there.

`$XDG_DATA_DIRS` could be added easily - but it is specified to fall back to `/usr/local/share:/usr/share` if undefined, so we might still want a configuration time switch to not use those default.
Looks like `zsh` always adds `/usr/local/share/zsh/site-functions` to `$fpath`, that works as well.

## TODOs:
- [x] User-visible changes noted in CHANGELOG.md

zanchey commented on CHANGELOG.md:
```diff
@@ -146,6 +146,7 @@
 - The `INTERNAL_WCWIDTH` build option has been removed, as fish now always uses an internal `wcwidth` function. It has a number of configuration options that make it more suitable for general use (#5777).
 - mandoc can now be used to format the output from `--help` if `nroff` is not installed, reducing the number of external dependencies on systems with `mandoc` installed (#5489).
 - Some bugs preventing building on Solaris-derived systems such as Illumos were fixed (#5458, #5461, #5611).
+- Vendored configuration like completions and functions are sourced first from `{/usr/local,/usr}/share/fish/vendor_{completions,conf,functions}.d` and second from `$PREFIX/share/fish/vendor_{completions,conf,functions}.d` . Previously, only the latter was used. The former paths can be customized using the CMake variable `EXTRA_VENDOR_PATH`; set it to an empty string to restore the old behavior ( ...
```
Rather than "vendored", this should be "Third-party" or similar... though really it's probably fifth-party!

krobelus commented:
I got rid of the new cmake variable and changed it to use `$XDG_DATA_DIRS`
instead, I think that's better.

Also I changed it to not unconditionally include `$PREFIX/share/fish/vendor_completions.d`, but rather `$__extra_completionsdir` (which defaults to the former) like before.

> * the default really should be `/usr/local/share/fish:$DATADIR`, respecting the prefix rather than always including `/usr`
> * using `XDG_DATA_DIRS` as discussed in #5029 is definitely a better approach, though again I'd try and respect the installation prefix

The XDG basedir spec says:
> If `$XDG_DATA_DIRS` is either not set or empty, a value equal to `/usr/local/share/:/usr/share/` should be used.

Should we still only use `/usr/local/share`?

I am deviating already on the "or empty" case, in that case my implementation adds no paths from XDG_DATADIRS (since it's an empty list), not sure if that's smart.

> * need to update the pkg-config interface as well

Does this need to know all paths? As I understand it's for packaging
systems to know where to drop third party scripts, that should be alright
as is. If we use `XDG_DATA_DIRS`, it's impossible to know all paths at
configuration time.

krobelus commented:
I have several packages on my system that put their fish configuration in `share/fish/functions` instead of the `share/fish/vendor_functions.d` so their packaging would need to be fixed for them to leverage the extra vendor directories.

faho commented:
On my system $XDG_DATA_DIRS has been set to also include some flatpak directories. It's set to $HOME/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share:/usr/local/share:/usr/share.

I *think* that's okay?

zanchey commented:
> On my system $XDG_DATA_DIRS has been set to also include some flatpak directories.

You'd hope that was only inside flatpak applications...

zanchey commented:
My apologies for how long this is taking - it takes a long time for me to get all the components lined up in my head.

I think this is close, but the default paths still need to respect `$PREFIX` because not all systems implement the XDG path spec, and may use different paths. So, if `XDG_DATA_DIRS` is not set, the default should be the current "extra" directory (defaulting to `/usr/local/share`) followed by `$DATADIR` (from CMake); perhaps deduplicating for cases where they are the same. On systems with the prefix set to `/usr`, which is all FHS/XDG-compliant systems, this will meet the spec.

I think this strikes the right balance between respecting the spec and behaving correctly in the real world.

I'm happy to take a crack at implementing this if you don't have time.

krobelus commented:
> So, if `XDG_DATA_DIRS` is not set, the default should be the current "extra" directory (defaulting to `/usr/local/share`) followed by `$DATADIR` (from CMake); perhaps deduplicating for cases where they are the same. On systems with the prefix set to `/usr`, which is all FHS/XDG-compliant systems, this will meet the spec.

Yeah, that makes sense, I did that now. So with a default PREFIX (`/usr/local`) and undefined XDG_DATA_DIRS, this results in only loading vendor conf from `/usr/local/share/fish/vendor_conf.d`.
And if `extra_confdir` is customized, then it's `${extra_confdir}:/usr/local/share/fish/vendor_conf.d`

Sadly, there seem to be many packages that drop their fish functions in `/usr/share/fish/functions` instead of `vendor_functions.d` (ex.: the file list for [fzf](https://www.archlinux.org/packages/community/x86_64/fzf/)). Which makes a fish installation not fully relocatable; I guess I'll go and file bug reports.

zanchey commented:
(The documentation also needs updating)

krobelus commented on share/config.fish:
```diff
@@ -66,17 +66,47 @@ if test -f $__fish_data_dir/__fish_build_paths.fish
+    set xdg_data_dirs $__fish_data_dir
+    # The extra directories default to /usr/local/share/*
+    # We need to have them take precedence over $__fish_data_dir
+    set extra_mode --prepend
```
Second thought --- I wonder if the additional complexity with append/prepend is necessary.
The intention was to fix the order to be like `/usr/local/share:/usr/share`.
At the moment this will use `${extra_completionsdir}:$__fish_datadir/vendor_completions.d`, both are in `/usr/local/share` by default. So why not swap them, and have the "extra" dirs always come last.

So far there has only been one directory called `vendor_completions.d` - the `extra_completionsdir` (that is created at installation time).

zanchey replied:
Yes, that makes sense.
